image/save: set a stable timestamp for assets

stevvooe opened the pull request:
When saving a docker image with `docker save`, output may have the current timestamp, resulting in slightly changed content each time the `save` command gets run. This patch attemtps to stabilize that effort to clean up some spots where we've missed setting the timestamps.

It's not totally clear that setting these timestamps to 0 is the correct behavior but it will fix the hash stability problem on output.

- Closes https://github.com/docker/cli/issues/5515
- (hopefully) Closes https://github.com/moby/moby/issues/49068

We still need some better testing for this and might need some discussion around _which_ timestamp we actually use.

stevvooe committed:
image/save: set a stable timestamp for assets

When saving a docker image with `docker save`, output may have the
current timestamp, resulting in slightly changed content each time the
`save` command gets run. This patch attemtps to stabilize that effort to
clean up some spots where we've missed setting the timestamps.

It's not totally clear that setting these timestamps to 0 is the correct
behavior but it will fix the hash stability problem on output.

Signed-off-by: Stephen Day <stephen.day@docker.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on image/tarexport/save.go:
```diff
@@ -576,12 +586,10 @@ func (s *saveSession) saveConfigAndLayer(ctx context.Context, id layer.ChainID,
-				return distribution.Descriptor{}, errors.Wrap(err, "could not set layer timestamp")
-			}
+	for _, fname := range []string{outDir, layerPath} {
+		// todo: maybe save layer created timestamp?
```
This TODO was added back in 2015 by @tonistiigi in https://github.com/moby/moby/commit/01ba0a935b9227583ec05a96e85273cc2276bb93#diff-9ac2cf7826d0cfc8316ca97dceae47447f431bc510fd4a4622ebeb53cc74d08aR295

I wonder if it's still clear what it's about, or if we should just drop it 🤔

stevvooe replied:
Not quite sure. This just does the `nil` check once but shouldn't change the behavior here. This PR still defers on a timestamp methodology, other than fixing the issue with the particular assets identified in https://github.com/docker/cli/issues/5515.

The other approach here would be to do a full walk and set them all to the same value after assembling the `save` file.

thaJeztah replied:
Yeah, was mostly wondering if we had a specific reason to preserve the layer timestamp, because if not, we might as well remove the TODO comment here.

Resetting to the image's timestamp (what I saw you did in this PR) makes sense to me; at least it makes things stable for the same image

thaJeztah commented:
Did a quick rebase of this PR; I'll have a look at Pawel's comments about referencing the upstream source; I'm considering moving the copied code to a separate file to match https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/os/path.go;l=19, which can make comparing with upstream easier, and to add a comment about the reference taken as starting point.

vvoland commented:
@stevvooe still interested in addressing the comments? Otherwise me or @thaJeztah  can carry it

thaJeztah commented on image/tarexport/save.go:
```diff
@@ -608,22 +616,82 @@ func (s *saveSession) saveConfig(legacyImg image.V1Image, outDir string, created
+	}
+
+	if err := system.Chtimes(path, atime, mtime); err != nil {
+		return fmt.Errorf("applying atime=%v and mtime=%v to %s: %w",
+			atime, mtime, path, err)
```
Just noticed that the `to <Path>` may be redundant here, because `system.Chtimes` uses `os.Chtimes`, which returns a `PathError` already; 

https://github.com/golang/go/blob/39f2032c17516a55c165d329dd5e2e07f49132b0/src/os/file_posix.go#L190-L192

```go
if e := syscall.UtimesNano(fixLongPath(name), utimes[0:]); e != nil {
	return &PathError{Op: "chtimes", Path: name, Err: e}
}
```

https://github.com/golang/go/blob/39f2032c17516a55c165d329dd5e2e07f49132b0/src/io/fs/fs.go#L257

```go
func (e *PathError) Error() string { return e.Op + " " + e.Path + ": " + e.Err.Error() }
```

thaJeztah commented:
> @stevvooe still interested in addressing the comments? Otherwise me or @thaJeztah can carry it

I had a branch locally with a "WIP" commit to move the code to a separate file; I can push (keeping it as a separate commit for now, but can squash)

thaJeztah commented:
@vvoland pushed; let me know if it looks good to you then I'll squash the commits

vvoland commented on image/tarexport/os_path.go:
```diff
@@ -0,0 +1,75 @@
+// license that can be found in the LICENSE file.
+
+// Code in this file is a modified version of go stdlib;
+// https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/os/path.go;l=19
```
Would be nice if we split this into two separate commits:

1. Import the function 1:1 from the go source
2. Adjust the code with our modifications

Like https://github.com/moby/moby/commit/56bb143a4d55f73a1e2ede192d782369df14abe5 + https://github.com/moby/moby/commit/5841ed4e5e0ea0a9b78fa6759cd9b5929463c313

thaJeztah replied:
Not a bad suggestion, it makes it easier indeed to see "what did we change". With the small note that of course the first copy would need _slight_ changes (we can't use the `filepathlite` as it's internal (but `filepath.VolumeName` is the same); https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/os/path.go;l=47

And need to add the `os.` before exported functions.

I'll work on that 👍

vvoland replied:
I don't think we necessarily need to adjust the first copy. While it makes the first commit not buildable, I think it's not a big problem. In practice it would only be a problem when bisecting, but since `git bisect skip` exists... Personally I wouldn't worry much about it.

thaJeztah replied:
OK; updated 👍 

(CI failure is a current incident with Docker Hub)

vvoland replied:
The changes to `image/tarexport/os_path.go` in 17a0fc5589d3d69f0cf8d7c3b4916847585f5ed0 should also go into a separate commit after the non-buildable one. So in future it can be cherry-picked on its own if the upstream code changes and we need to rebase our forked copy.

thaJeztah replied:
Ah! Done 👍

thaJeztah committed:
image/tarexport: add fork of os.MkdirAll (non-buildable commit)

Fork the os.MkdirAll function, so that we can make changes to add support
for setting atime and mtime. This commit contains a 1:1 fork, but won't
build; code is taken from:
https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/os/path.go;l=19-66

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
image/tarexport: patch MkDirall to accept atime, mtime

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
